File: scala/Optics.scala
/**
 * Created by ASUS-PC on 2016/9/6.
 */
import math._
import scala.collection.mutable._

/**
 *
 * @param epsilon The maximum distance (radius) to consider
 * @param minPts The number of points required to form a cluster
 * @param points data points
 */
class Optics(epsilon:Double,
             minPts:Long,
             points:Array[Vector[Double]]) {

  val UNDEFINED = Double.MaxValue
  val NOISE = Int.MaxValue
  
  class OpticsPoint(_id:Int) {
    val id = _id
    var neighbors = Array.empty[Int]
    var processed = false
    var coreDist = UNDEFINED
    var reachDist = UNDEFINED
    var clusterId = NOISE

    def getPoint: Vector[Double] = points(id)
  }
  
  val optPoints  = points.indices.map(new OpticsPoint(_))

  def getOptPoint(p:Vector[Double]): OpticsPoint = optPoints(points.indexOf(p))

  def getNeighbors(p:OpticsPoint): Array[OpticsPoint] = p.neighbors.map(optPoints(_))

  def getDistance(X1:Vector[Double], X2:Vector[Double]): Double = sqrt(X1.zip(X2).map(x => pow(x._1 - x._2, 2)).sum)
  def getDistance(X1:OpticsPoint, X2:OpticsPoint): Double = getDistance(X1.getPoint,X2.getPoint)

  def initPoint(p:OpticsPoint):Unit = {
    val distances = optPoints.map(x => getDistance(x, p))
    val neighbors = optPoints.zip(distances).filter( np => (np._2 < epsilon) && (np._1.id !=p.id) ).sortBy(_._2)
    p.neighbors = neighbors.map(_._1.id).toArray
    p.coreDist = if (neighbors.length >= minPts) neighbors(minPts.toInt - 1)._2 else UNDEFINED
  }

  def enqueue(Seeds: ArrayBuffer[(OpticsPoint, Double)], p: OpticsPoint, newD: Double): Unit ={
    var i = 0
    while ( i<Seeds.length && Seeds(i)._2<newD) { i+=1 }
    Seeds.insert(i, (p,newD))
  }

  def moveUp(Seeds: ArrayBuffer[(OpticsPoint, Double)], p: OpticsPoint, newD: Double): Unit ={
    val i = Seeds.map(_._1).indexOf(p)
    Seeds.remove(i)
    enqueue(Seeds, p, newD)
  }

  def dequeue(Seeds: ArrayBuffer[(OpticsPoint, Double)]): (OpticsPoint, Double) ={
    Seeds.remove(0)
  }

  def update(Seeds:ArrayBuffer[(OpticsPoint, Double)], N: Array[OpticsPoint], p:OpticsPoint): Unit = {
//    println("updating seeds of point:" + p.id.toString + ", neighborhood:(" + N.map(_.id.toString).mkString(",") + ")")
    for (o <- N) {
      if (!o.processed) {
        val newReachDist = max(p.coreDist, getDistance(p, o))
        if (o.reachDist == UNDEFINED) {
          o.reachDist = newReachDist
          enqueue(Seeds, o, newReachDist)
        } else if ( newReachDist < o.reachDist) {
          o.reachDist = newReachDist
          moveUp(Seeds, o, newReachDist)
        }
      }
    }
  }

  def extractCluster(orderedPoints: ArrayBuffer[OpticsPoint]): ArrayBuffer[OpticsPoint] ={
    var C = 0
    for (p <- orderedPoints){
      if (p.reachDist > epsilon) {
        if (p.coreDist <= epsilon) {
          C += 1
          p.clusterId = C
        } else p.clusterId = NOISE
      } else {
        p.clusterId = C
      }
    }
    orderedPoints
  }

  def run(): ArrayBuffer[OpticsPoint] ={
    val orderedPoints = new ArrayBuffer[OpticsPoint]
    optPoints.foreach(initPoint)
    val Seeds = ArrayBuffer.empty[(OpticsPoint,Double)]
    for (p <- optPoints) {
      if (!p.processed) {
        val N = getNeighbors(p)
        p.processed = true
        orderedPoints.append(p)
        if ( p.coreDist != UNDEFINED) {
          update(Seeds, N, p)
          while ( Seeds.nonEmpty){
            val seed = dequeue(Seeds)
            val q = seed._1
            val NN = getNeighbors(q)
            q.processed = true
            orderedPoints.append(q)
            if ( q.coreDist != UNDEFINED ) {
              update(Seeds, NN, q)
            }
          }
        }
      }
    }
    extractCluster(orderedPoints)
  }
}

File: python/optics.py
import numpy
import sys
from math import *

UNDEFINED = sys.float_info.max
NOISE = sys.maxsize


class OpticsPoint():
    def __init__(self, id):
        self.pid = id
        self.neighbors = numpy.empty(0, dtype=int)
        self.processed = False
        self.coreDist = UNDEFINED
        self.reachDist = UNDEFINED
        self.clusterId = NOISE


class Optics:

    def __init__(self,
                 epsilon:   float,
                 minPts:    int,
                 points:    numpy.array,
                 distType:  str):
        self.epsilon = epsilon
        self.minPts = minPts
        self.points = points
        self.distType = distType
        self.optPoints = numpy.array([OpticsPoint(i) for i in range(len(points))])

    def getPoint(self, p:OpticsPoint):
        return self.points[p.pid]

    def getNeighbors(self, p:OpticsPoint):
        return [self.optPoints[i] for i in p.neighbors]

    def getDistance(self, p:OpticsPoint, o:OpticsPoint):
        X1 = self.getPoint(p)
        X2 = self.getPoint(o)
        if self.distType == 'euclidean':
            sub = numpy.subtract(X1, X2)
            sqr = numpy.square(sub)
            return sqrt(sqr.sum())
        elif self.distType == 'angle':
            absX1 = sqrt(numpy.square(X1).sum())
            absX2 = sqrt(numpy.square(X2).sum())
            dotProd = numpy.prod([X1, X2], axis=0).sum()
            return dotProd/(absX1 * absX2)

    def initPoint(self, p:OpticsPoint):
        neighbors = []
        distances = []
        for o in self.optPoints:
            if o != p:
                distance = self.getDistance(p, o)
                if distance < self.epsilon:
                    neighbors.append(o.pid)
                    distances.append(distance)
        p.neighbors = numpy.array(neighbors)
        if len(p.neighbors) >= self.minPts:
            p.coreDist = sorted(distances)[self.minPts-1]

    def enqueue(self, Seeds, p:OpticsPoint, newD):
        i = 0
        while (i < len(Seeds)) and (Seeds[i][1] < newD):
            i += 1
        Seeds.insert(i, (p, newD))

    def moveUp(self, Seeds, p:OpticsPoint, newD):
        i = 0
        while i < len(Seeds):
            if Seeds[i][0] == p:
                Seeds.pop(i)
            i += 1
        self.enqueue(Seeds, p, newD)

    def dequeue(self, Seeds):
        return Seeds.pop(0)

    def update(self, Seeds, N, p:OpticsPoint):
        for o in N:
            if not o.processed:
                newReachDist = max(p.coreDist, self.getDistance(p, o))
                if o.reachDist == UNDEFINED:
                    o.reachDist = newReachDist
                    self.enqueue(Seeds, o, newReachDist)
                elif newReachDist < o.reachDist:
                    o.reachDist = newReachDist
                    self.moveUp(Seeds, o, newReachDist)

    def extractCluster(self, orderedPoints):
        C = 0
        for p in orderedPoints:
            if p.reachDist > self.epsilon:
                if p.coreDist <= self.epsilon:
                    C += 1
                    p.clusterId = C
            else:
                p.clusterId = C
        return orderedPoints

    def run(self):
        for p in self.optPoints:
            self.initPoint(p)
        orderedPoints = []
        Seeds = []
        for p in self.optPoints:
            if not p.processed:
                N = self.getNeighbors(p)
                p.processed = True
                orderedPoints.append(p)
                if p.coreDist != UNDEFINED:
                    self.update(Seeds, N, p)
                    while len(Seeds) > 0:
                        seed = self.dequeue(Seeds)
                        q = seed[0]
                        NN = self.getNeighbors(q)
                        q.processed = True
                        orderedPoints.append(q)
                        if q.coreDist != UNDEFINED:
                            self.update(Seeds, NN, q)
        return self.extractCluster(orderedPoints)
